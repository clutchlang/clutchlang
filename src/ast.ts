export * from './language/ast/token';
