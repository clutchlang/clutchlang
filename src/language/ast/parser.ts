import * as tokens from './token';

/**
 * Represents an unresolved node in the abstract syntax tree.
 */
export abstract class AstNode {
  public abstract accept<R, C>(
    visitor: AstVisitor<R, C>,
    context?: C
  ): R | undefined;

  /**
   * First token that represents this AST node.
   */
  public abstract get first(): tokens.Token;

  /**
   * Last token that represents this AST node.
   */
  public abstract get last(): tokens.Token;
}

/**
 * Represents a simple AST node that originates from a single token.
 */
export abstract class SimpleNode extends AstNode {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an operator node parsed from a single token.
 */
export class Operator extends SimpleNode {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitOperator(this, context);
  }
}

/**
 * Represents a marker type of AST node for all statements.
 */
export abstract class Statement extends AstNode {}

/**
 * Represents a marker type of AST node for all expressions.
 */
export abstract class Expression extends Statement {}

/**
 * Represents a simple AST node that is also an expression.
 */
export abstract class SimpleExpression extends Expression {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an expression that utilizes an operator.
 */
export abstract class OperatorExpression extends Expression {
  constructor(
    public readonly operator: Operator,
    public readonly target: Expression
  ) {
    super();
  }
}

/**
 * Represents an expression in the form of `<OP><EXPR>`.
 */
export class PrefixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPrefixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.operator.first;
  }

  public get last(): tokens.Token {
    return this.target.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP>`.
 */
export class PostfixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPostfixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.operator.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP><EXPR>`.
 */
export class BinaryExpression extends OperatorExpression {
  constructor(
    operator: Operator,
    public readonly left: Expression,
    public readonly right: Expression
  ) {
    super(operator, left);
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitBinaryExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.left.first;
  }

  public get last(): tokens.Token {
    return this.right.last;
  }
}

/**
 * Represents an expression in the form of `<TARGET>.<PROPERTY>`.
 */
export class PropertyExpression extends Expression {
  constructor(
    public readonly target: Expression,
    public readonly property: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPropertyExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.property.last;
  }
}

/**
 * Represents a list of arguments passed to a function call.
 */
export class ArgumentList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly args: Expression[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitArgumentList(this, context);
  }
}

/**
 * Represents an expression for a function call.
 *
 * May _optionally_ contain a @member target.
 */
export class CallExpression extends Expression {
  constructor(
    public readonly name: Identifier,
    public readonly args: ArgumentList,
    public readonly target?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitCallExpression(this, context);
  }

  public get first(): tokens.Token {
    return (this.target || this.name).first;
  }

  public get last(): tokens.Token {
    return this.args.last;
  }
}

export class ConditionalExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly condition: Expression,
    public readonly body: Expression | StatementBlock,
    public readonly elseCondition?: Expression,
    public readonly elseBody?: Expression | StatementBlock
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitConditionalExpression(this, context);
  }

  public get last(): tokens.Token {
    return (this.elseBody || this.body).last;
  }
}

export class GroupExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression: Expression,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitGroupExpression(this, context);
  }
}

/**
 * Represents an expression for a literal identifier (name).
 */
export class Identifier extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitIdentifier(this, context);
  }

  public get name(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal boolean.
 */
export class LiteralBoolean extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralBoolean(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal number.
 */
export class LiteralNumber extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralNumber(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal string.
 */
export class LiteralString extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralString(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents a variable declared anywhere in scope.
 */
export class VariableDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly type?: Identifier,
    public readonly value?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitVariableDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return (this.value || this.type || this.name).last;
  }
}

/**
 * Represents a list of parameters for a function or closure definition.
 */
export class ParameterList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly params: VariableDeclaration[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitParameterList(this, context);
  }
}

/**
 * Represents a series of statements.
 */
export class StatementBlock extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly statements: Statement[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitStatementBlock(this, context);
  }
}

/**
 * Represents a return statement.
 */
export class ReturnStatement extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitReturnStatement(this, context);
  }

  public get last(): tokens.Token {
    return this.expression ? this.expression.last : this.first;
  }
}

/**
 * Represents a function declared anywhere in scope.
 */
export class FunctionDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly params: ParameterList,
    public readonly body: Expression | StatementBlock,
    public readonly returnType?: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitFunctionDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return this.body.last;
  }
}

/**
 * Represents a type declared anywhere in scope.
 */
export class TypeDeclaration extends AstNode {
  constructor(
    public readonly modifiers: tokens.Token[],
    public readonly name: Identifier,
    public readonly members: Array<FunctionDeclaration | VariableDeclaration>,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitTypeDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.modifiers.length ? this.modifiers[0] : this.name.first;
  }
}

/**
 *
 */
export class ModuleRoot extends AstNode {
  constructor(
    public readonly declarations: Array<
      FunctionDeclaration | TypeDeclaration | VariableDeclaration
    >
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitModuleRoot(this, context);
  }

  public get first(): tokens.Token {
    return this.declarations[0].first;
  }

  public get last(): tokens.Token {
    return this.declarations[this.declarations.length - 1].last;
  }
}

export class AstParser {}

/**
 * An AST visitor that by default does nothing visiting an AST node.
 *
 * It is intended to be a superclass for classes that use the visitor pattern
 * primarily as a dispatch mechanism (and do not need to recursively need to
 * visit the whole structure) and only need to visit a small number of node
 * types.
 */
export class AstVisitor<R, C> {
  public visitArgumentList(_: ArgumentList, __?: C): R | undefined {
    return;
  }

  public visitBinaryExpression(_: BinaryExpression, __?: C): R | undefined {
    return;
  }

  public visitCallExpression(_: CallExpression, __?: C): R | undefined {
    return;
  }

  public visitConditionalExpression(
    _: ConditionalExpression,
    __?: C
  ): R | undefined {
    return;
  }

  public visitFunctionDeclaration(
    _: FunctionDeclaration,
    __?: C
  ): R | undefined {
    return;
  }

  public visitGroupExpression(_: GroupExpression, __?: C): R | undefined {
    return;
  }

  public visitIdentifier(_: Identifier, __?: C): R | undefined {
    return;
  }

  public visitLiteralBoolean(_: LiteralBoolean, __?: C): R | undefined {
    return;
  }

  public visitLiteralNumber(_: LiteralNumber, __?: C): R | undefined {
    return;
  }

  public visitLiteralString(_: LiteralString, __?: C): R | undefined {
    return;
  }

  public visitModuleRoot(_: ModuleRoot, __?: C): R | undefined {
    return;
  }

  public visitOperator(_: Operator, __?: C): R | undefined {
    return;
  }

  public visitParameterList(_: ParameterList, __?: C): R | undefined {
    return;
  }

  public visitPrefixExpression(_: PrefixExpression, __?: C): R | undefined {
    return;
  }

  public visitPostfixExpression(_: PostfixExpression, __?: C): R | undefined {
    return;
  }

  public visitPropertyExpression(_: PropertyExpression, __?: C): R | undefined {
    return;
  }

  public visitReturnStatement(_: ReturnStatement, __?: C): R | undefined {
    return;
  }

  public visitStatementBlock(_: StatementBlock, __?: C): R | undefined {
    return;
  }

  public visitTypeDeclaration(_: TypeDeclaration, __?: C): R | undefined {
    return;
  }

  public visitVariableDeclaration(
    _: VariableDeclaration,
    __?: C
  ): R | undefined {
    return;
  }
}
