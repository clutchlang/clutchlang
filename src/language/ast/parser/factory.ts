import { Token } from '../lexer';
import * as ast from './nodes';

/**
 * Indirection interface for creating AST nodes.
 */
export class AstFactory {
  public createOperator<T extends ast.OperatorType>(
    token: Token,
    type: T
  ): ast.Operator<T> {
    return new ast.Operator(token, type);
  }

  public createPrefixExpression(
    operator: ast.Operator<ast.PrefixOperatorType>,
    target: ast.Expression
  ): ast.PrefixExpression {
    return new ast.PrefixExpression(operator, target);
  }

  public createPostfixExpression(
    operator: ast.Operator<ast.PostfixOperatorType>,
    target: ast.Expression
  ): ast.PostfixExpression {
    return new ast.PostfixExpression(operator, target);
  }

  public createBinaryExpression(
    operator: ast.Operator<ast.BinaryOperatorType>,
    left: ast.Expression,
    right: ast.Expression
  ): ast.BinaryExpression {
    return new ast.BinaryExpression(operator, left, right);
  }

  public createPropertyExpression<T extends ast.Expression>(
    target: T,
    property: ast.Identifier
  ): ast.PropertyExpression<T> {
    return new ast.PropertyExpression(target, property);
  }

  public createCallExpression<E extends ast.Expression>(
    target: E,
    args: ast.Expression[],
    last: Token
  ): ast.CallExpression<E> {
    return new ast.CallExpression(target, args, last);
  }

  public createConditionalExpression(
    first: Token,
    condition: ast.Expression,
    body: ast.Expression | ast.StatementBlock,
    elseBody?: ast.Expression | ast.StatementBlock
  ): ast.ConditionalExpression {
    return new ast.ConditionalExpression(first, condition, body, elseBody);
  }

  public createGroupExpression<E extends ast.Expression>(
    first: Token,
    expression: E,
    last: Token
  ): ast.GroupExpression<E> {
    return new ast.GroupExpression(first, expression, last);
  }

  public createIdentifier(token: Token): ast.Identifier {
    return new ast.Identifier(token);
  }

  public createLiteralBoolean(token: Token): ast.LiteralBoolean {
    return new ast.LiteralBoolean(token);
  }

  public createLiteralNumber(token: Token): ast.LiteralNumber {
    return new ast.LiteralNumber(token);
  }

  public createLiteralString(token: Token): ast.LiteralString {
    return new ast.LiteralString(token);
  }

  public createVariableDeclaration(
    name: ast.Identifier,
    isConst: boolean,
    type?: ast.Identifier,
    value?: ast.Expression
  ): ast.VariableDeclaration {
    return new ast.VariableDeclaration(name, isConst, type, value);
  }

  public createParameterList(
    first: Token,
    params: ast.VariableDeclaration[],
    last: Token
  ): ast.ParameterList {
    return new ast.ParameterList(first, params, last);
  }

  public createStatementBlock(
    first: Token,
    statements: ast.Statement[],
    last: Token
  ): ast.StatementBlock {
    return new ast.StatementBlock(first, statements, last);
  }

  public createReturnStatement(
    first: Token,
    expression?: ast.Expression
  ): ast.ReturnStatement {
    return new ast.ReturnStatement(first, expression);
  }

  public createFunctionDeclaration(
    name: ast.Identifier,
    params?: ast.ParameterList,
    body?: ast.Expression | ast.StatementBlock,
    returnType?: ast.Identifier
  ): ast.FunctionDeclaration {
    return new ast.FunctionDeclaration(name, params, body, returnType);
  }

  public createModuleDeclaration(
    declarations: Array<
      ast.FunctionDeclaration | ast.TypeDeclaration | ast.VariableDeclaration
    >
  ): ast.ModuleDeclaration {
    return new ast.ModuleDeclaration(declarations);
  }
}
