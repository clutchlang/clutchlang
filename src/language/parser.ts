export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
