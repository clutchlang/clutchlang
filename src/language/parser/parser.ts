import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
