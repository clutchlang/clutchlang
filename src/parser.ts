export { Codes } from './parser/codes';
export { Lexer } from './parser/lexer';
export { Token } from './parser/token';
export { Tokens } from './parser/tokens';
