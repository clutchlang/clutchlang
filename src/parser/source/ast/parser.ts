import {
  RegExpToken,
  StringToken,
  SymbolToken,
  Token,
  TokenKind,
} from '../tokenizer/tokens';
import {
  AstCompilationUnit,
  AstExpression,
  AstFunctionDeclaration,
  AstInvocationExpression,
  AstLiteralBoolean,
  AstLiteralIdentifier,
  AstLiteralNumber,
  AstLiteralString,
  AstParenthesizedExpression,
} from './node';
import { TokenScanner } from './scanner';

export class AstParser {
  constructor(private readonly scanner: TokenScanner) {}

  public parseCompilationUnit(): AstCompilationUnit {
    const results: AstFunctionDeclaration[] = [];
    while (!this.scanner.isDone) {
      // TODO: Loosen once alternatives are supported.
      this.scanRequired(RegExpToken.Identifier);
      results.push(this.parseFunctionDeclaration());
    }
    return new AstCompilationUnit(results);
  }

  /**
   * Returns a parsed expression, or undefined if there is none.
   */
  private parseExpression(): AstExpression | undefined {
    const token = this.scanner.peek();
    switch (token.kind) {
      case SymbolToken.LParen:
        this.scanRequired(SymbolToken.LParen);
        const expressions = this.parseExpressions();
        this.scanRequired(SymbolToken.RParen);
        return new AstParenthesizedExpression(expressions);
      case RegExpToken.Identifier:
        if (this.scanner.peek(1).kind === SymbolToken.LParen) {
          const identifier = this.scanner.read();
          this.scanRequired(SymbolToken.LParen);
          const invocation = new AstInvocationExpression(
            new AstLiteralIdentifier(identifier),
            this.parseExpressions()
          );
          this.scanRequired(SymbolToken.RParen);
          return invocation;
        }
        return new AstLiteralIdentifier(this.scanner.read());
      case RegExpToken.LiteralBoolean:
        return new AstLiteralBoolean(this.scanner.read());
      case RegExpToken.LiteralNumber:
        return new AstLiteralNumber(this.scanner.read());
      case RegExpToken.LiteralString:
        return new AstLiteralString(this.scanner.read());
      default:
        return undefined;
    }
  }

  private parseFunctionDeclaration(): AstFunctionDeclaration {
    const identifier = this.scanner.lastMatch[0];
    this.scanRequired(StringToken.Arrow);
    if (this.scanOptional(SymbolToken.LCurly)) {
      const fn = new AstFunctionDeclaration(
        identifier,
        this.parseExpressions()
      );
      this.scanRequired(SymbolToken.RCurly);
      return fn;
    }
    const expression = this.parseExpression()!;
    return new AstFunctionDeclaration(identifier, [expression]);
  }

  private parseExpressions(): AstExpression[] {
    const expressions: AstExpression[] = [];
    let expression = this.parseExpression();
    while (expression) {
      expressions.push(expression);
      expression = this.parseExpression();
    }
    return expressions;
  }

  private scanOptional(token: TokenKind): Token | undefined {
    if (this.scanner.scan(token)) {
      return this.scanner.lastMatch[0];
    }
    return undefined;
  }

  private scanRequired(token: TokenKind): void {
    /* istanbul ignore next */
    if (!this.scanner.scan(token)) {
      // The lexer should prevent this happening in a typical program.
      /* istanbul ignore next */
      const next = this.scanner.peek().kind;
      /* istanbul ignore next */
      throw new SyntaxError(`Expected ${token.name} got ${next.name}.`);
    }
  }
}
