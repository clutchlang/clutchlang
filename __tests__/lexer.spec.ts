import { Token } from 'moo';
import { lexer } from '../src/parser';

function tokenize(program: string): Token[] {
  lexer.reset(program);
  const results: Token[] = [];
  for (const token of lexer) {
    results.push(token);
  }
  return results;
}

test('should tokenize a simple program', () => {
  lexer.reset('main => {\n' + '}\n');
  expect(tokenize('main => {}')).toMatchSnapshot();
});
