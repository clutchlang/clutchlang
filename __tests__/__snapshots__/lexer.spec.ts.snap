// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should tokenize a blank program 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "main",
    "toString": [Function],
    "type": "identifier",
    "value": "main",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "=>",
    "toString": [Function],
    "type": "arrow",
    "value": "=>",
  },
  Object {
    "col": 8,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": "{",
    "toString": [Function],
    "type": "lcurly",
    "value": "{",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "}",
    "toString": [Function],
    "type": "rcurly",
    "value": "}",
  },
]
`;

exports[`should tokenize a function that returns true 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "check",
    "toString": [Function],
    "type": "identifier",
    "value": "check",
  },
  Object {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "=>",
    "toString": [Function],
    "type": "arrow",
    "value": "=>",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "true",
    "toString": [Function],
    "type": "boolean",
    "value": "true",
  },
]
`;

exports[`should tokenize a function with parameters and a return type 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "capitalize",
    "toString": [Function],
    "type": "identifier",
    "value": "capitalize",
  },
  Object {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "(",
    "toString": [Function],
    "type": "lparen",
    "value": "(",
  },
  Object {
    "col": 13,
    "line": 1,
    "lineBreaks": 0,
    "offset": 12,
    "text": "word",
    "toString": [Function],
    "type": "identifier",
    "value": "word",
  },
  Object {
    "col": 17,
    "line": 1,
    "lineBreaks": 0,
    "offset": 16,
    "text": ":",
    "toString": [Function],
    "type": "colon",
    "value": ":",
  },
  Object {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 19,
    "line": 1,
    "lineBreaks": 0,
    "offset": 18,
    "text": "string",
    "toString": [Function],
    "type": "identifier",
    "value": "string",
  },
  Object {
    "col": 25,
    "line": 1,
    "lineBreaks": 0,
    "offset": 24,
    "text": ")",
    "toString": [Function],
    "type": "rparen",
    "value": ")",
  },
  Object {
    "col": 26,
    "line": 1,
    "lineBreaks": 0,
    "offset": 25,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 27,
    "line": 1,
    "lineBreaks": 0,
    "offset": 26,
    "text": "=>",
    "toString": [Function],
    "type": "arrow",
    "value": "=>",
  },
  Object {
    "col": 29,
    "line": 1,
    "lineBreaks": 0,
    "offset": 28,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 30,
    "line": 1,
    "lineBreaks": 0,
    "offset": 29,
    "text": "string",
    "toString": [Function],
    "type": "identifier",
    "value": "string",
  },
  Object {
    "col": 36,
    "line": 1,
    "lineBreaks": 0,
    "offset": 35,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 37,
    "line": 1,
    "lineBreaks": 0,
    "offset": 36,
    "text": "{",
    "toString": [Function],
    "type": "lcurly",
    "value": "{",
  },
  Object {
    "col": 38,
    "line": 1,
    "lineBreaks": 0,
    "offset": 37,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 39,
    "line": 1,
    "lineBreaks": 0,
    "offset": 38,
    "text": "return",
    "toString": [Function],
    "type": "keyword",
    "value": "return",
  },
  Object {
    "col": 45,
    "line": 1,
    "lineBreaks": 0,
    "offset": 44,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 46,
    "line": 1,
    "lineBreaks": 0,
    "offset": 45,
    "text": "word",
    "toString": [Function],
    "type": "identifier",
    "value": "word",
  },
  Object {
    "col": 50,
    "line": 1,
    "lineBreaks": 0,
    "offset": 49,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 51,
    "line": 1,
    "lineBreaks": 0,
    "offset": 50,
    "text": "}",
    "toString": [Function],
    "type": "rcurly",
    "value": "}",
  },
]
`;

exports[`should tokenize a variable assignment 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "let",
    "toString": [Function],
    "type": "keyword",
    "value": "let",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "name",
    "toString": [Function],
    "type": "identifier",
    "value": "name",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "=",
    "toString": [Function],
    "type": "assign",
    "value": "=",
  },
  Object {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "'Matan'",
    "toString": [Function],
    "type": "string",
    "value": "'Matan'",
  },
]
`;

exports[`should tokenize invocation 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "cat",
    "toString": [Function],
    "type": "identifier",
    "value": "cat",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": ".",
    "toString": [Function],
    "type": "dot",
    "value": ".",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "meow",
    "toString": [Function],
    "type": "identifier",
    "value": "meow",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": "(",
    "toString": [Function],
    "type": "lparen",
    "value": "(",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": ")",
    "toString": [Function],
    "type": "rparen",
    "value": ")",
  },
]
`;

exports[`should tokenize two top-level functions 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "one",
    "toString": [Function],
    "type": "identifier",
    "value": "one",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "=>",
    "toString": [Function],
    "type": "arrow",
    "value": "=>",
  },
  Object {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 8,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "1",
    "toString": [Function],
    "type": "number",
    "value": "1",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 1,
    "offset": 8,
    "text": "
",
    "toString": [Function],
    "type": "linebreak",
    "value": "
",
  },
  Object {
    "col": 1,
    "line": 2,
    "lineBreaks": 0,
    "offset": 9,
    "text": "two",
    "toString": [Function],
    "type": "identifier",
    "value": "two",
  },
  Object {
    "col": 4,
    "line": 2,
    "lineBreaks": 0,
    "offset": 12,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 5,
    "line": 2,
    "lineBreaks": 0,
    "offset": 13,
    "text": "=>",
    "toString": [Function],
    "type": "arrow",
    "value": "=>",
  },
  Object {
    "col": 7,
    "line": 2,
    "lineBreaks": 0,
    "offset": 15,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 8,
    "line": 2,
    "lineBreaks": 0,
    "offset": 16,
    "text": "2",
    "toString": [Function],
    "type": "number",
    "value": "2",
  },
]
`;
